GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,465 @@
+            0.0_dp, vec_UT_psi, num_x_nodes)
+
+       ! 3) (V*Sigma^-1) * (U^T*psi)
+       call dgemm('N', 'N', num_points, 1, num_x_nodes, 1.0_dp, mat_VT_s, num_points, vec_UT_psi, &
+            num_x_nodes, 0.0_dp, weights_work, num_points)
+
+       weights(i_point, :) = weights_work(:)
+
+       ! calculate the maximum error of the fitting
+       call calculate_max_error(num_points, tau_points, omega_points, weights_work, num_x_nodes, x_mu, &
+            psi, current_point, max_error, transformation_type)
+    end do ! i_point
+
+    deallocate (x_mu, psi, mat_A, weights_work, vec_S, mat_U, mat_VT, work, iwork, mat_VT_s, vec_UT_psi)
+
+  end subroutine get_transformation_weights
+
+  !> \brief Calculate the auxiliary matrix for cosine/sin transformation for tau to omega or viceversa
+  !! @param[in] num_points: Number of mesh points.
+  !! @param[in] tau_points: imaginary time grid points
+  !! @param[in] omega_points: imaginary frequency grid points
+  !! @param[in] num_x_nodes: Number of node in the interval [e_min,e_max]
+  !! @param[in] x_mu : Transition energy (nodes in the interval [e_min,e_max])
+  !! @param[inout] psi: corresponding auxiliary function (see transformation type definition)
+  !! @param[inout] mat_A: auxiliary matrix (see transformation type definition)  
+  !! @param[in] i_point: pointer for the current grid point
+  !! @param[inout] current_point:  current grid point ether omega(i_point) or tau_(i_point)
+  !! @param[in] transformation type :
+  !!        (1) the cosine transform cos(it) -> cos(iw): psi(omega,x), mat_A = cos(omega*tau)*psi(tau,x) 
+  !!        (2) the cosine transform cos(iw) -> cos(it): psi(tau,x)  , mat_A = cos(omega*tau)*psi(omega,x)
+  !!        (3) the sine transform   sin(it) -> sin(iw): psi(omega,x), mat_A = sin(omega*tau)*psi(tau,x)
+  subroutine calculate_psi_and_mat_A(num_points, tau_points, omega_points, num_x_nodes, x_mu, psi, &
+       mat_A, i_point, current_point, transformation_type)
+
+    integer, intent(in)                                :: num_points, num_x_nodes, i_point
+    real(kind=dp), allocatable, dimension(:), &
+         intent(in)                                    :: tau_points, omega_points, x_mu
+    real(kind=dp), allocatable, dimension(:), &          
+         intent(inout)                                 :: psi         
+    real(kind=dp), allocatable, dimension(:, :), &
+         intent(inout)                                 :: mat_A
+    real(kind=dp), intent(inout)                       :: current_point
+    integer, intent(in)                                :: transformation_type
+
+    ! Internal variables
+    integer                                            :: i_node, j_point
+    real(kind=dp)                                      :: tau, omega
+
+    ! Begin work
+
+    ! the cosine transform cos(it) -> cos(iw)
+    if (transformation_type.eq.1) then
```

Comment:
Same deal here as with the error quantification. Same code, different points and psi definition. Just write once and inject the correct definitions depending on the transformation type. If you're interested, I can post some snippets or pair code it

Reply:
We can do it for later developing.